x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/gap/interval_utils.ts

```diff
@@ -0,0 +1,225 @@
+    }
+    return merged;
+  }, [] as Interval[]);
+};
+
+/**
+ * Subtracts a list of intervals from a single base interval, returning any remaining parts that do not overlap.
+ *
+ * @param base - The base interval.
+ * @param intervalsToSubtract - Intervals to remove from the base.
+ * @returns An array of intervals representing the remainder of the base interval.
+ *
+ * @example
+ * const base: Interval = { gte: new Date('2023-10-19T12:00:00Z'), lte: new Date('2023-10-19T13:00:00Z') };
+ * const toSubtract: Interval[] = [{ gte: new Date('2023-10-19T12:20:00Z'), lte: new Date('2023-10-19T12:30:00Z') }];
+ * // Result splits the base into two: 12:00-12:20 and 12:30-13:00
+ * const result = subtractIntervals(base, toSubtract);
+ * // result = [
+ * //   { gte: 2023-10-19T12:00:00Z, lte: 2023-10-19T12:20:00Z },
+ * //   { gte: 2023-10-19T12:30:00Z, lte: 2023-10-19T13:00:00Z }
+ * // ]
+ */
+export const subtractIntervals = (base: Interval, intervalsToSubtract: Interval[]): Interval[] => {
+  let result: Interval[] = [base];
+  for (const toSubtract of intervalsToSubtract) {
+    result = result.flatMap((current) => {
+      const overlap = getOverlap(current, toSubtract);
+      if (!overlap) return [current];
+
+      const remainder: Interval[] = [];
+      const currentStart = current.gte.getTime();
+      const currentEnd = current.lte.getTime();
+      const overlapStart = overlap.gte.getTime();
+      const overlapEnd = overlap.lte.getTime();
+
+      if (overlapStart > currentStart) {
+        remainder.push({ gte: new Date(currentStart), lte: new Date(overlapStart) });
+      }
+
+      if (overlapEnd < currentEnd) {
+        remainder.push({ gte: new Date(overlapEnd), lte: new Date(currentEnd) });
+      }
+
+      return remainder;
+    });
+  }
+  return result;
+};
+
+/**
+ * Applies subtractIntervals to each interval in an array and merges the result,
+ * effectively removing intervalsToSubtract from all given intervals.
+ *
+ * @param intervals - The original intervals from which to subtract.
+ * @param intervalsToSubtract - The intervals to remove.
+ * @returns An array of intervals representing what remains after all subtractions.
+ *
+ * @example
+ * const intervals: Interval[] = [
+ *   { gte: new Date('2023-10-19T12:00:00Z'), lte: new Date('2023-10-19T12:30:00Z') },
+ *   { gte: new Date('2023-10-19T12:40:00Z'), lte: new Date('2023-10-19T13:00:00Z') }
+ * ];
+ * const toSubtract: Interval[] = [{ gte: new Date('2023-10-19T12:20:00Z'), lte: new Date('2023-10-19T12:45:00Z') }];
+ * // For the first interval: removing 12:20-12:30 leaves 12:00-12:20
+ * // For the second interval: removing 12:20-12:45 overlaps partially, leaving 12:45-13:00
+ * // After merging, result:
+ * // [
+ * //   { gte: 2023-10-19T12:00:00Z, lte: 2023-10-19T12:20:00Z },
+ * //   { gte: 2023-10-19T12:45:00Z, lte: 2023-10-19T13:00:00Z }
+ * // ]
+ * const result = subtractAllIntervals(intervals, toSubtract);
+ */
+export const subtractAllIntervals = (
```

Comment:
With `m` intervals and `n` intervalsToSubtract, we'll have to do `m*n` comparisons. We can reduce this to `m+n`, if we can assume that both `intervals` and `intervalsToSubtract` are sorted, by stepping through the two arrays together. There are some edge cases to deal with that make it trickier, but we should check the performance of this function with the max number of intervals we expect and consider optimizing it.